src/program.fs

```diff
@@ -88,19 +106,38 @@ module Program =
         { program
             with onError = onError }
 
+#if FABLE
+    type private CachedValue<'a>() =
+        [<DefaultValue>] val mutable Value:'a
+        member this.SetValue(generator: unit ->'a) =
+            if weakEqual this.Value null then
+                this.Value <- generator ()
+#else
+    type private CachedValue<'a>() =
+        let lockObject = obj()
+        [<DefaultValue>] val mutable Value:'a
+        member this.SetValue(generator: unit ->'a) =
+            lock lockObject (fun () ->
+                if (obj.ReferenceEquals(this.Value, Unchecked.defaultof<'a>)) then
+                    this.Value <- generator ()
+            )
+#endif
+
     /// Start the program loop.
     /// arg: argument to pass to the init() function.
     /// program: program created with 'mkSimple' or 'mkProgram'.
     let runWith (arg: 'arg) (program: Program<'arg, 'model, 'msg, 'view>) =
         let (model,cmd) = program.init arg
+        let setStateCache = CachedValue<'model -> unit>()
```

Comment:
Not sure if it's related, but it'd be interesting to see if using a cache can help fix https://github.com/fable-elmish/react/issues/12

Reply:
I don't think that [controlled components](https://reactjs.org/docs/forms.html#controlled-components) can be used with `requestAnimationFrame` (or anything async) ever.

They currently work in Elmish because `MailboxProcessor` is executing `Post` synchronously, if it was doing int any other way (By using a `setTimeout` for example) it would break them.

[ReactControlledComponent](https://github.com/facebook/react/blob/master/packages/events/ReactControlledComponent.js#L15) is in charge of restoring changes after any change event

What happens in such cases:

Textbox: `Hello |orld` (`|` is the cursor)
➡️ User type `w`
➡️ OnChange (queue a `requestAnimationFrame`)
➡️ Browser set Textbox: `Hello w|orld`
➡️ After running event handler diff DOM, see that the text in the box is `Hello world` instead of `Hello orld` and correct (Cursor move to end)
➡️ Animation frame run and set `value = Hello world`
➡️ React re-render and set the value of the DOM textbox to `Hello world` (Cursor move to end again)

Reply:
(Yes i'm not a fan of the fact that `requestAnimationFrame` version is the "default" one currently as I feel that it's something that has a heavy risk of tripping users trying to use something that is very common in react land)

Reply:
How would you handle application with a lot of update coming at more than "60fps" rate without the `requestAnimationFrame` ? Would this be a problem ?

Reply:
I fail to see your point.

The range of applications that need anything that update so fast is small (games & visual demo mainly) and here it break a few of the core features of React.

I wouldn't be against it being the default if it didn't break things. The majority of frontend applications aren't games they are form input+display ones and having something not adapted to that presented as the "default" and used everywhere, in every tutorial out there seem counterproductive.

* As someone that know react (Or is following a tutorial) I want to use `Fable.Elmish.React` for my TODO list app.
* I use the template that everyone use in fable and follow react best practices (or my react tutorial)
* My app behave weirdly
* By creating an issue on github (Or angry twittering, or angry slacking) I discover that controlled components don't work well, but there is a solution to use `DefaultValue` and `Ref` an make a non controlled one (With all the problems it can create)
* Later I Include an awesome component that does auto completion with a very slick UI. And boom, component is using controlled components inside.
* Same issue/twitter/slack route I learn about `withReactUnoptimized `

Repeat for all future users of `Fable.Elmish.React` (Except maybe the few that both do games and don't use forms)

Essentially i'd love to see `withReactUnoptimized` become `withReact` and `withReact` to become `withReactAnimationFrameOptimized` that would have a comment listing the consequences.

Reply:
Thanks for the explanations. 

After thinking about the problem I got in mind, I more or less consider users could use `requestAnimationFrame` to dispatch a message periodically if needed.